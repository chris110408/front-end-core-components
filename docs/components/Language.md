# Language

The Language component provides global localization state via React's [context API](https://reactjs.org/docs/context.html). It has support built in for:

* URL based language in the form of `/<lang>/path/to/resource/`
* Multiple languages
* Alternative languages redirecting to a main language e.g. `en-US` -> `en`
* Persisting to local storage
* Accepting languages via `window.navigator.locale` when supported

Each of the above features are optional.

## Dependencies

* react `^16.3`
* prop-types `^15.6.2`

## Usage

*The following examples assume the usage of the Gatsby framework.*

### Provider

In order to maintain an accurate URL state it is recommended to use a library such as `@reach/router` with the `Location` component. Other libraries can be used so long as the values for `location.pathname`, `location.search` and `location.hash` are kept up to date. If the prop `useURL` is false then `location` is not required.

```javascript
import React from 'react'
import { Location } from '@reach/router'
import { navigate } from 'gatsby'
import { Language } from '@input-output-hk/front-end-core-components'

const App = () => (
  <Location>
    {({ location }) => (
      <Language.Provider
        location={{
          // location is optional when the useURL prop is false
          pathname: location.pathname,
          search: location.search,
          hash: location.hash
        }}
        availableLanguages={[
          {
            key: 'en',
            label: 'English US',
            // flag is optional
            flag: 'ðŸ‡ºðŸ‡¸',
            locale: 'en-US'
          },
          {
            key: 'fr',
            label: 'French FR',
            flag: 'â€ŽðŸ‡«ðŸ‡·',
            locale: 'fr-FR'
          }
        ]}
        alternativeLanguages={[
          // Alternative languages are optional
          {
            key: 'en-GB',
            languageKey: 'en'
          },
          {
            key: 'fr-CA',
            languageKey: 'fr'
          }
        ]}
        onUpdate={({ lang, prevLang, url, prevURL }) => {
          // Optional callback whenever the language is updated
          // When the component mounts `prevLang` and `prevURL` will be null
          // An alternative to using this prop would be to use the consumer
          // and useEffect(() => { ... }, [ lang ]) where appropriate

          // When useURL is false the `url` and `prevURL` arguments are undefined

          // Examples
          // * track event to analytics
          // * navigate to the new URL
          // * display an on screen confirmation
          if (prevURL && prevURL !== url) navigate(url)
        }}
        persistLang={
          true // defaults to true, whether to persist language to local storage or not
        }
        useNavigator={
          true // defaults to true, whether to attempt to grab language from window.navigator.language or not
        }
        useURL={
          true // defaults to true, whether to use the URL from the given location or not
        }
      >
        <MyComponent />
      </Language.Provider>
    )}
  </Location>
)

```

### Consumer

```javascript
import React from 'react'
import { Language } from '@input-output-hk/front-end-core-components'

const MyComponent = () => (
  <Language.Consumer>
    {({
      lang,
      setLang,
      locale,
      flag,
      label,
      availableLanguages,
      alternativeLanguages
    }) => (
      <div>
        <select value={key} onChange={e => setLang(e.target.value)}>
          {availableLanguages.map(({ key, label, flag, locale }) => (
            <option key={key} value={key}>
              {flag} {label} ({locale})
            </option>
          ))}
        </select>
        <p>Selected language, {flag} {label}</p>
      </div>
    )}
  </Language.Consumer>
)

```
