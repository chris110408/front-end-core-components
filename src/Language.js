import React, { createContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'

const LanguageContext = createContext()
const Consumer = LanguageContext.Consumer

const Provider = ({
  children,
  location: { pathname, search, hash } = {},
  availableLanguages,
  alternativeLanguages = [],
  onUpdate = ({ lang, prevLang, url, prevURL }) => {},
  persistLang = true,
  useNavigator = true,
  useURL = true
}) => {
  const [ lang, setLanguage ] = useState(resolveLanguage(getDefaultLanguage()))

  /**
   * Persists language key to local storage when persistLang is true
   *
   * @param {String} language The language to persist to local storage
   * @returns {Void}
   */
  function persistToLocalStorage (language) {
    if (persistLang) window && window.localStorage && window.localStorage.setItem('lang', language)
  }

  /**
   * Fetches an initial default language going through all possible sources as follows
   *   - url
   *   - local storage
   *   - navigator
   *   - first language in availableLanguages
   *
   * @returns {String} Language key
   */
  function getDefaultLanguage () {
    const urlLang = useURL ? getURLLang() : ''
    // Use URL language where available
    if (isValidLanguage(urlLang)) return urlLang
    // Fallback to local storage or default config
    const localStorageLang = (persistLang && (window && window.localStorage && window.localStorage.getItem('lang'))) || ''
    // Users browser language
    const navigatorLanguage = ((useNavigator && (window && window.navigator && window.navigator.language)) || '').substring(0, 2)
    // Use local storage first as they have already visited the site
    if (isValidLanguage(localStorageLang)) return localStorageLang
    // Use the users browser language if it supported by the site
    if (isValidLanguage(navigatorLanguage)) return navigatorLanguage
    // Fall back to the first available language in the config
    return availableLanguages[0].key
  }

  /**
   * Resolves any potential language from alternativeLanguages to availableLanguages
   *
   * @param {String} lang Language key
   * @returns {String} Language key
   */
  function resolveLanguage (lang) {
    const filteredAlternativeLanguages = alternativeLanguages.filter(({ key }) => key === lang)
    if (filteredAlternativeLanguages.length > 0) return filteredAlternativeLanguages.shift().languageKey
    return lang
  }

  /**
   * Fetches the language object for a specific key from availableLanguages
   * or resolves from alternativeLanguages
   *
   * @param {String} language Language key
   * @returns {<Object|undefined>}
   */
  function getLanguage (language) {
    const lang = availableLanguages.filter(({ key }) => key === language).shift()
    if (!lang) language = (alternativeLanguages.filter(({ key }) => key === language).shift() || {}).languageKey
    return lang || availableLanguages.filter(({ key }) => key === language).shift()
  }

  /**
   * Checks is a language key is valid, i.e. is valid from availableLanguages
   * or alternativeLanguages
   *
   * @param {String} language Language key
   * @returns {Boolean}
   */
  function isValidLanguage (language) {
    return Boolean(getLanguage(language))
  }

  /**
   * Fetches the possible language from the URL
   *
   * @returns {String}
   */
  function getURLLang () {
    return pathname.split('/').splice(1)[0] || ''
  }

  /**
   * Checks if the language from the URL is a valid language
   *
   * @returns {Boolean}
   */
  function languageSetInURL () {
    return isValidLanguage(getURLLang())
  }

  /**
   * Builds the URL for a particular language
   *
   * @param {String} language Language key
   * @returns {String}
   */
  function getURL (language) {
    const pathParts = pathname.split('/').splice(1)
    if (languageSetInURL()) {
      pathParts.splice(0, 1, language)
    } else {
      pathParts.unshift(language)
    }

    return `/${pathParts.filter(part => Boolean(part)).join('/').replace(/\/{2,}/g, '/')}/${search}${hash}`
  }

  useEffect(() => {
    persistToLocalStorage(lang)
    const updateArgs = {
      lang,
      prevLang: null
    }

    if (useURL) {
      updateArgs.url = getURL(lang)
      updateArgs.prevURL = null
    }

    onUpdate(updateArgs)
  }, [])

  /**
   * Runs through all steps when updating the language
   *
   * @param {String} language Language key
   * @returns {Void}
   */
  function setLang (language) {
    if (!isValidLanguage(language)) throw new Error(`Invalid language ${language}`)
    persistToLocalStorage(language)

    const updateArgs = {
      lang: language,
      prevLang: lang
    }

    if (useURL) {
      updateArgs.url = getURL(language)
      updateArgs.prevURL = `${pathname}${search}${hash}`
    }

    onUpdate(updateArgs)
    setLanguage(language)
  }

  return (
    <LanguageContext.Provider
      value={{
        key: lang,
        setLang,
        locale: getLanguage(lang).locale,
        flag: getLanguage(lang).flag,
        label: getLanguage(lang).label,
        availableLanguages,
        alternativeLanguages
      }}
    >
      {children}
    </LanguageContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.node.isRequired,
  location: (props, propName, componentName) => {
    const requiredKeys = [ 'pathname', 'search', 'hash' ]
    if (props.useURL && (!props[propName] || typeof props[propName] !== 'object' || Array.isArray(props[propName]))) {
      return new Error(`When 'useURL' is true, location must be provided as an object with location.pathname, location.search and location.hash`)
    } else if (props.useURL) {
      while (requiredKeys.length > 0) {
        const key = requiredKeys.shift()
        if (typeof props[propName][key] !== 'string') return new Error(`location.${key} must be a string`)
      }
    }
  },
  availableLanguages: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    flag: PropTypes.string,
    locale: PropTypes.string.isRequired
  })).isRequired,
  alternativeLanguages: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string.isRequired,
    languageKey: PropTypes.string.isRequired
  })),
  onUpdate: PropTypes.func,
  persistLang: PropTypes.bool,
  useNavigator: PropTypes.bool,
  useURL: PropTypes.bool
}

export {
  Provider,
  Consumer
}
