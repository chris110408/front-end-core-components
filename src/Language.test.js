import React from 'react'
import { Provider, Consumer } from './Language'
import { mount, shallow } from 'enzyme'
import { act } from 'react-dom/test-utils'

const DEFAULT_PROVIDER_PROPS = {
  location: {
    pathname: '/',
    search: '',
    hash: ''
  },
  availableLanguages: [
    {
      key: 'en',
      locale: 'en-US',
      label: 'US English',
      flag: 'â€ŽðŸ‡ºðŸ‡¸'
    },
    {
      key: 'fr',
      locale: 'fr-FR',
      'label': 'French',
      flag: 'â€ŽðŸ‡«ðŸ‡·'
    }
  ],
  children: (<p>Language provider children</p>)
}

function getProviderWrapper ({ props = {}, renderFunction }) {
  return renderFunction(
    <Provider
      location={props.location === null ? undefined : { ...DEFAULT_PROVIDER_PROPS.location, ...(props.location || {}) }}
      availableLanguages={props.availableLanguages || DEFAULT_PROVIDER_PROPS.availableLanguages}
      alternativeLanguages={props.alternativeLanguages}
      onUpdate={props.onUpdate}
      persistLang={props.persistLang}
      useNavigator={props.useNavigator}
      useURL={props.useURL}
    >
      {props.children ? props.children : DEFAULT_PROVIDER_PROPS.children}
    </Provider>
  )
}

describe('Language context', () => {
  describe('<Provider />', () => {
    let provider
    beforeEach(() => {
      provider = getProviderWrapper({ renderFunction: shallow })
    })

    test('it renders the provider component correctly', () => {
      expect(provider).toMatchSnapshot()
    })

    describe('when the provider is mounted', () => {
      let onUpdate
      beforeEach(() => {
        onUpdate = jest.fn()
        provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate } })
      })

      test('the language is persisted to local storage', () => {
        expect(window.localStorage.setItem).toBeCalledWith('lang', 'en')
      })

      test('onUpdate is called with the correct arguments', () => {
        expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null, url: '/en/', prevURL: null })
      })

      describe('when there is a URL path, search and hash', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          provider = getProviderWrapper({
            renderFunction: mount,
            props: {
              onUpdate,
              location: { pathname: '/path/to/resource/', search: '?query=some%20search%20term', hash: '#id' }
            }
          })
        })

        test('onUpdate is called with the correct arguments', () => {
          expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null, url: '/en/path/to/resource/?query=some%20search%20term#id', prevURL: null })
        })
      })

      describe('when persistLang is false', () => {
        beforeEach(() => {
          window.localStorage.setItem.mockReset()
          provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate, persistLang: false } })
        })

        test('the language is not persisted to local storage', () => {
          expect(window.localStorage.setItem).not.toBeCalled()
        })
      })

      describe('when useURL is false', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate, useURL: false, location: null } })
        })

        test('onUpdate is called with the correct arguments', () => {
          expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null })
        })

        describe('when location is still used', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate, useURL: false, location: { pathname: '/fr/' } } })
          })

          test('onUpdate is called with the correct arguments', () => {
            expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null })
          })
        })
      })

      describe('when the language is set in the URL initially', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          provider = getProviderWrapper({
            renderFunction: mount,
            props: {
              onUpdate,
              location: { pathname: '/fr/path/to/resource/' }
            }
          })
        })

        test('onUpdate is called with the correct arguments', () => {
          expect(onUpdate).toBeCalledWith({ lang: 'fr', prevLang: null, url: '/fr/path/to/resource/', prevURL: null })
        })

        describe('when a valid alternative language is used', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            provider = getProviderWrapper({
              renderFunction: mount,
              props: {
                onUpdate,
                alternativeLanguages: [
                  {
                    key: 'fr-CA',
                    languageKey: 'fr'
                  }
                ],
                location: { pathname: '/fr-CA/path/to/resource/' }
              }
            })
          })

          test('onUpdate is called with the correct arguments', () => {
            expect(onUpdate).toBeCalledWith({ lang: 'fr', prevLang: null, url: '/fr/path/to/resource/', prevURL: null })
          })
        })
      })

      describe('when the language is set in local storage initially', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          window.localStorage.getItem.mockImplementation((key) => key === 'lang' ? 'fr' : undefined)
          provider = getProviderWrapper({
            renderFunction: mount,
            props: {
              onUpdate
            }
          })
        })

        afterEach(() => {
          window.localStorage.getItem.mockReset()
        })

        test('onUpdate is called with the correct arguments', () => {
          expect(onUpdate).toBeCalledWith({ lang: 'fr', prevLang: null, url: '/fr/', prevURL: null })
        })

        describe('when the language in local storage is invalid', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            window.localStorage.getItem.mockImplementation((key) => key === 'lang' ? 'de' : undefined)
            provider = getProviderWrapper({
              renderFunction: mount,
              props: {
                onUpdate
              }
            })
          })

          test('onUpdate is called with the correct arguments', () => {
            expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null, url: '/en/', prevURL: null })
          })
        })
      })

      describe('when the language is set from the navigator initially', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          window.navigator.language = 'fr-FR'
          provider = getProviderWrapper({
            renderFunction: mount,
            props: {
              onUpdate
            }
          })
        })

        afterEach(() => {
          window.navigator.language = 'de-DE'
        })

        test('onUpdate is called with the correct arguments', () => {
          expect(onUpdate).toBeCalledWith({ lang: 'fr', prevLang: null, url: '/fr/', prevURL: null })
        })

        describe('when useNavigator is false', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            provider = getProviderWrapper({
              renderFunction: mount,
              props: {
                onUpdate,
                useNavigator: false
              }
            })
          })

          test('onUpdate is called with the correct arguments', () => {
            expect(onUpdate).toBeCalledWith({ lang: 'en', prevLang: null, url: '/en/', prevURL: null })
          })
        })
      })
    })
  })

  describe('<Consumer />', () => {
    let onUpdate, wrapper
    beforeEach(() => {
      onUpdate = jest.fn()
      wrapper = getProviderWrapper({
        renderFunction: mount,
        props: {
          onUpdate,
          children: (
            <Consumer>
              {({ lang, setLang, locale, flag, label, availableLanguages }) => (
                <div
                  id='consumer-value-holder'
                  consumervalue={{
                    lang,
                    setLang,
                    locale,
                    flag,
                    label,
                    availableLanguages
                  }}
                />
              )}
            </Consumer>
          )
        }
      })
    })

    afterEach(() => {
      onUpdate.mockReset()
    })

    test('the consumer values are correct', () => {
      expect(wrapper.find('#consumer-value-holder').prop('consumervalue')).toMatchSnapshot()
    })

    describe('when setLang is called', () => {
      beforeEach(() => {
        act(() => {
          wrapper.find('#consumer-value-holder').prop('consumervalue').setLang('fr')
        })
        wrapper.update()
      })

      test('the new language is persisted to local storage', () => {
        expect(window.localStorage.setItem).toBeCalledWith('lang', 'fr')
      })

      test('the consumer values are correctly updated', () => {
        expect(wrapper.find('#consumer-value-holder').prop('consumervalue')).toMatchSnapshot()
      })

      test('when attempting to set an invalid language', () => {
        expect(() => {
          wrapper.find('#consumer-value-holder').prop('consumervalue').setLang('zh-CN')
        }).toThrow('Invalid language zh-CN')
      })
    })
  })
})
