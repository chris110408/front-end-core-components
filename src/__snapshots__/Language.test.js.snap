// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language context <Consumer /> the consumer values are correct 1`] = `
Object {
  "availableLanguages": Array [
    Object {
      "flag": "â€ŽðŸ‡ºðŸ‡¸",
      "key": "en",
      "label": "US English",
      "locale": "en-US",
    },
    Object {
      "flag": "â€ŽðŸ‡«ðŸ‡·",
      "key": "fr",
      "label": "French",
      "locale": "fr-FR",
    },
  ],
  "flag": "â€ŽðŸ‡ºðŸ‡¸",
  "label": "US English",
  "lang": "en",
  "locale": "en-US",
  "setLang": [Function],
}
`;

exports[`Language context <Consumer /> when setLang is called the consumer values are correctly updated 1`] = `
Object {
  "availableLanguages": Array [
    Object {
      "flag": "â€ŽðŸ‡ºðŸ‡¸",
      "key": "en",
      "label": "US English",
      "locale": "en-US",
    },
    Object {
      "flag": "â€ŽðŸ‡«ðŸ‡·",
      "key": "fr",
      "label": "French",
      "locale": "fr-FR",
    },
  ],
  "flag": "â€ŽðŸ‡«ðŸ‡·",
  "label": "French",
  "lang": "fr",
  "locale": "fr-FR",
  "setLang": [Function],
}
`;

exports[`Language context <Provider /> it renders the provider component correctly 1`] = `
<ContextProvider
  value={
    Object {
      "alternativeLanguages": Array [],
      "availableLanguages": Array [
        Object {
          "flag": "â€ŽðŸ‡ºðŸ‡¸",
          "key": "en",
          "label": "US English",
          "locale": "en-US",
        },
        Object {
          "flag": "â€ŽðŸ‡«ðŸ‡·",
          "key": "fr",
          "label": "French",
          "locale": "fr-FR",
        },
      ],
      "flag": "â€ŽðŸ‡ºðŸ‡¸",
      "label": "US English",
      "lang": "en",
      "locale": "en-US",
      "setLang": [Function],
    }
  }
>
  <p>
    Language provider children
  </p>
</ContextProvider>
`;
